{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyP4k5JcLqgB/fH51QzzMq6f"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"BZBngLBTdAFz","executionInfo":{"status":"ok","timestamp":1675615488618,"user_tz":180,"elapsed":8347,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"fa188209-774b-47f5-dd42-37ced72cea43"},"outputs":[{"output_type":"stream","name":"stdout","text":["Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/\n","Collecting biopython\n","  Downloading biopython-1.80-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.1 MB)\n","\u001b[2K     \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[32m3.1/3.1 MB\u001b[0m \u001b[31m29.8 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n","\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.8/dist-packages (from biopython) (1.21.6)\n","Installing collected packages: biopython\n","Successfully installed biopython-1.80\n"]}],"source":["!pip install biopython --upgrade"]},{"cell_type":"code","source":["# https://biopython.org/docs/1.76/api/Bio.Entrez.html\n","from Bio import Entrez\n","Entrez.email = 'taniagmangolini@gmail.com'"],"metadata":{"id":"_D1Bldb2d1s7","colab":{"base_uri":"https://localhost:8080/"},"executionInfo":{"status":"ok","timestamp":1675616147486,"user_tz":180,"elapsed":280,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"b2e3232f-56ae-4f32-f769-bcd82beda4f9"},"execution_count":2,"outputs":[{"output_type":"stream","name":"stdout","text":["dict_keys(['Database', 'Query', 'CorrectedQuery', 'SpelledQuery'])\n"]}]},{"cell_type":"markdown","source":["## Databases"],"metadata":{"id":"r1t5o5pPr7au"}},{"cell_type":"code","source":["handle = Entrez.einfo()\n","record = Entrez.read(handle)\n","record[\"DbList\"]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"rQnivtyir4ys","executionInfo":{"status":"ok","timestamp":1675619220299,"user_tz":180,"elapsed":262,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"bf31e25b-1814-4896-e287-03e8ca8172a5"},"execution_count":69,"outputs":[{"output_type":"execute_result","data":{"text/plain":["['pubmed', 'protein', 'nuccore', 'ipg', 'nucleotide', 'structure', 'genome', 'annotinfo', 'assembly', 'bioproject', 'biosample', 'blastdbinfo', 'books', 'cdd', 'clinvar', 'gap', 'gapplus', 'grasp', 'dbvar', 'gene', 'gds', 'geoprofiles', 'homologene', 'medgen', 'mesh', 'nlmcatalog', 'omim', 'orgtrack', 'pmc', 'popset', 'proteinclusters', 'pcassay', 'protfam', 'pccompound', 'pcsubstance', 'seqannot', 'snp', 'sra', 'taxonomy', 'biocollections', 'gtr']"]},"metadata":{},"execution_count":69}]},{"cell_type":"markdown","source":["## Espell function"],"metadata":{"id":"W3tmyTSjg_yn"}},{"cell_type":"code","source":["records = Entrez.read(Entrez.espell(db='pmc', term='biobbython'))\n","for key, sugestions in records.items():\n","  print(f'{key} : {sugestions}')"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"DSkE6ttdg-q2","executionInfo":{"status":"ok","timestamp":1675616481378,"user_tz":180,"elapsed":428,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"c40078a4-2d9b-44c2-a1b0-113859317165"},"execution_count":7,"outputs":[{"output_type":"stream","name":"stdout","text":["Database : pmc\n","Query : biobbython\n","CorrectedQuery : biopython\n","SpelledQuery : ['', 'biopython']\n"]}]},{"cell_type":"markdown","source":["## Esearch function\n","Permite fazer consultas no NCBI."],"metadata":{"id":"SldcgIW5hs5R"}},{"cell_type":"code","source":["records = Entrez.read(Entrez.esearch(db='pmc', term='biopython'))\n","print(records)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"osh3hrcMh0O9","executionInfo":{"status":"ok","timestamp":1675616638741,"user_tz":180,"elapsed":323,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"9cecc498-9302-4345-82a4-0f871066418f"},"execution_count":8,"outputs":[{"output_type":"stream","name":"stdout","text":["{'Count': '2715', 'RetMax': '20', 'RetStart': '0', 'IdList': ['9898917', '9883684', '9886943', '9896449', '9894515', '9882952', '9877618', '9876871', '9872083', '9710682', '9837158', '9864840', '9861041', '9855806', '9850167', '9845338', '9826903', '9835423', '9837567', '9836094'], 'TranslationSet': [], 'TranslationStack': [{'Term': 'biopython[All Fields]', 'Field': 'All Fields', 'Count': '2733', 'Explode': 'N'}, 'GROUP'], 'QueryTranslation': 'biopython[All Fields]'}\n"]}]},{"cell_type":"code","source":["handle = Entrez.esearch(db=\"gene\",retmax=\"10\",term=\"P53 AND Homo Sapiens [organism]\")\n","record = Entrez.read(handle)\n","print(record)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"GToSQlbEq3R8","executionInfo":{"status":"ok","timestamp":1675618928931,"user_tz":180,"elapsed":273,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"99eb629e-634c-44e0-c15e-7ab43f1b573b"},"execution_count":63,"outputs":[{"output_type":"stream","name":"stdout","text":["{'Count': '2595', 'RetMax': '10', 'RetStart': '0', 'IdList': ['7157', '128385374', '1956', '348', '7124', '7422', '3569', '7040', '4524', '2064'], 'TranslationSet': [{'From': 'Homo Sapiens[organism]', 'To': '\"Homo sapiens\"[Organism]'}], 'TranslationStack': [{'Term': 'P53[All Fields]', 'Field': 'All Fields', 'Count': '24210', 'Explode': 'N'}, {'Term': '\"Homo sapiens\"[Organism]', 'Field': 'Organism', 'Count': '307665', 'Explode': 'Y'}, 'AND'], 'QueryTranslation': 'P53[All Fields] AND \"Homo sapiens\"[Organism]'}\n"]}]},{"cell_type":"code","source":["handle = Entrez.esearch(db=\"nucleotide\",term=\"Cypripedioideae[Orgn] AND matK[Gene]\")\n","record = Entrez.read(handle)\n","record[\"Count\"]"],"metadata":{"colab":{"base_uri":"https://localhost:8080/","height":35},"id":"lm8p_UIZsLm9","executionInfo":{"status":"ok","timestamp":1675619266844,"user_tz":180,"elapsed":446,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"f94c5ae4-5884-46c2-b707-6cee20cae1ec"},"execution_count":70,"outputs":[{"output_type":"execute_result","data":{"text/plain":["'830'"],"application/vnd.google.colaboratory.intrinsic+json":{"type":"string"}},"metadata":{},"execution_count":70}]},{"cell_type":"markdown","source":["## Esummary function\n","Permite buscar mais informações sobre um item.*texto em itálico*"],"metadata":{"id":"3J_RcgAHiwrD"}},{"cell_type":"code","source":["records = Entrez.read(Entrez.esearch(db='pmc', term='biopython'))\n","papers = records['IdList']\n","for paper in papers[:10]:\n","  summary = Entrez.read(Entrez.esummary(db='pmc', id=paper))[0]\n","  print(summary)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"MUF66m_0ii8R","executionInfo":{"status":"ok","timestamp":1675617100158,"user_tz":180,"elapsed":3757,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"1faf5cb5-3185-4cab-cb94-0ba062a133e4"},"execution_count":14,"outputs":[{"output_type":"stream","name":"stdout","text":["{'Item': [], 'Id': '9898917', 'PubDate': '2023 Feb 3', 'EPubDate': '2023 Feb 3', 'Source': 'BMC Biol', 'AuthorList': ['Sternke-Hoffmann R', 'Pauly T', 'Norrild RK', 'Hansen J', 'Tucholski F', 'Høie MH', 'Marcatili P', 'Dupré M', 'Duchateau M', 'Rey M', 'Malosse C', 'Metzger S', 'Boquoi A', 'Platten F', 'Egelhaaf SU', 'Chamot-Rooke J', 'Fenk R', 'Nagel-Steger L', 'Haas R', 'Buell AK'], 'Title': 'Widespread amyloidogenicity potential of multiple myeloma patient-derived immunoglobulin light chains', 'Volume': '21', 'Issue': '', 'Pages': '21', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '0', 'doi': '10.1186/s12915-022-01506-w', 'pmcid': 'PMC9898917'}, 'DOI': '10.1186/s12915-022-01506-w', 'FullJournalName': 'BMC Biology', 'SO': '2023 Feb 3;21:21'}\n","{'Item': [], 'Id': '9883684', 'PubDate': '2019 Aug 22', 'EPubDate': '2019 Aug 22', 'Source': 'Bioinformatics', 'AuthorList': ['Hugoson E', 'Lam WT', 'Guy L'], 'Title': 'miComplete: weighted quality evaluation of assembled microbial genomes', 'Volume': '36', 'Issue': '3', 'Pages': '936-937', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '31504158', 'doi': '10.1093/bioinformatics/btz664', 'pmcid': 'PMC9883684'}, 'DOI': '10.1093/bioinformatics/btz664', 'FullJournalName': 'Bioinformatics', 'SO': '2019 Aug 22;36(3):936-937'}\n","{'Item': [], 'Id': '9886943', 'PubDate': '2023 Jan 30', 'EPubDate': '2023 Jan 30', 'Source': 'Sci Rep', 'AuthorList': ['Kwon T', 'Hanschen ER', 'Hovde BT'], 'Title': 'Addressing the pervasive scarcity of structural annotation in eukaryotic algae', 'Volume': '13', 'Issue': '', 'Pages': '1687', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36717613', 'doi': '10.1038/s41598-023-27881-0', 'pmcid': 'PMC9886943'}, 'DOI': '10.1038/s41598-023-27881-0', 'FullJournalName': 'Scientific Reports', 'SO': '2023 Jan 30;13:1687'}\n","{'Item': [], 'Id': '9896449', 'PubDate': '2023 Feb 3', 'EPubDate': '2023 Feb 3', 'Source': 'Nat Commun', 'AuthorList': ['Chen WH', 'Hajduczki A', 'Martinez EJ', 'Bai H', 'Matz H', 'Hill TM', 'Lewitus E', 'Chang WC', 'Dawit L', 'Peterson CE', 'Rees PA', 'Ajayi AB', 'Golub ES', 'Swafford I', 'Dussupt V', 'David S', 'Mayer SV', 'Soman S', 'Kuklis C', 'Corbitt C', 'King J', 'Choe M', 'Sankhala RS', 'Thomas PV', 'Zemil M', 'Wieczorek L', 'Hart T', 'Duso D', 'Kummer L', 'Yan L', 'Sterling SL', 'Laing ED', 'Broder CC', 'Williams JK', 'Davidson E', 'Doranz BJ', 'Krebs SJ', 'Polonis VR', 'Paquin-Proulx D', 'Rolland M', 'Reiley WW', 'Gromowski GD', 'Modjarrad K', 'Dooley H', 'Joyce MG'], 'Title': 'Shark nanobodies with potent SARS-CoV-2 neutralizing activity and broad sarbecovirus reactivity', 'Volume': '14', 'Issue': '', 'Pages': '580', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '0', 'doi': '10.1038/s41467-023-36106-x', 'pmcid': 'PMC9896449'}, 'DOI': '10.1038/s41467-023-36106-x', 'FullJournalName': 'Nature Communications', 'SO': '2023 Feb 3;14:580'}\n","{'Item': [], 'Id': '9894515', 'PubDate': '2023 Feb 2', 'EPubDate': '2023 Feb 2', 'Source': 'Nat Commun', 'AuthorList': ['Normandin E', 'Rudy M', 'Barkas N', 'Schaffner SF', 'Levine Z', 'Padera RF Jr', 'Babadi M', 'Mukerji SS', 'Park DJ', 'MacInnis BL', 'Siddle KJ', 'Sabeti PC', 'Solomon IH'], 'Title': 'High-depth sequencing characterization of viral dynamics across tissues in fatal COVID-19 reveals compartmentalized infection', 'Volume': '14', 'Issue': '', 'Pages': '574', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36732505', 'doi': '10.1038/s41467-022-34256-y', 'pmcid': 'PMC9894515'}, 'DOI': '10.1038/s41467-022-34256-y', 'FullJournalName': 'Nature Communications', 'SO': '2023 Feb 2;14:574'}\n","{'Item': [], 'Id': '9882952', 'PubDate': '2023 Jan 27', 'EPubDate': '2023 Jan 27', 'Source': 'Comput Struct Biotechnol J', 'AuthorList': ['Jian X', 'Zhang Y', 'Zhao J', 'Zhao Z', 'Lu M', 'Xie L'], 'Title': 'CoV2-TCR: a web server for screening TCR CDR3 from TCR immune repertoire of COVID-19 patients and their recognized SARS-CoV-2 epitopes', 'Volume': '', 'Issue': '', 'Pages': '', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '0', 'doi': '10.1016/j.csbj.2023.01.038', 'pmcid': 'PMC9882952'}, 'DOI': '10.1016/j.csbj.2023.01.038', 'FullJournalName': 'Computational and Structural Biotechnology Journal', 'SO': '2023 Jan 27;'}\n","{'Item': [], 'Id': '9877618', 'PubDate': '2023 Jan 12', 'EPubDate': '2023 Jan 12', 'Source': 'Front Plant Sci', 'AuthorList': ['Ahmed B', 'Haque MA', 'Iquebal MA', 'Jaiswal S', 'Angadi UB', 'Kumar D', 'Rai A'], 'Title': 'DeepAProt: Deep learning based abiotic stress protein sequence classification and identification tool in cereals', 'Volume': '13', 'Issue': '', 'Pages': '1008756', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36714750', 'doi': '10.3389/fpls.2022.1008756', 'pmcid': 'PMC9877618'}, 'DOI': '10.3389/fpls.2022.1008756', 'FullJournalName': 'Frontiers in Plant Science', 'SO': '2023 Jan 12;13:1008756'}\n","{'Item': [], 'Id': '9876871', 'PubDate': '2022 Dec 3', 'EPubDate': '2022 Dec 3', 'Source': 'J Comput Aided Mol Des', 'AuthorList': ['Umedera K', 'Yoshimori A', 'Chen H', 'Kouji H', 'Nakamura H', 'Bajorath J'], 'Title': 'DeepCubist: Molecular Generator for Designing Peptidomimetics based on Complex three-dimensional scaffolds', 'Volume': '37', 'Issue': '2', 'Pages': '107-115', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36462089', 'doi': '10.1007/s10822-022-00493-y', 'pmcid': 'PMC9876871'}, 'DOI': '10.1007/s10822-022-00493-y', 'FullJournalName': 'Journal of Computer-Aided Molecular Design', 'SO': '2022 Dec 3;37(2):107-115'}\n","{'Item': [], 'Id': '9872083', 'PubDate': '2023 Jan 24', 'EPubDate': '2023 Jan 24', 'Source': 'Nat Commun', 'AuthorList': ['Parsons MF', 'Allan MF', 'Li S', 'Shepherd TR', 'Ratanalert S', 'Zhang K', 'Pullen KM', 'Chiu W', 'Rouskin S', 'Bathe M'], 'Title': '3D RNA-scaffolded wireframe origami', 'Volume': '14', 'Issue': '', 'Pages': '382', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36693871', 'doi': '10.1038/s41467-023-36156-1', 'pmcid': 'PMC9872083'}, 'DOI': '10.1038/s41467-023-36156-1', 'FullJournalName': 'Nature Communications', 'SO': '2023 Jan 24;14:382'}\n","{'Item': [], 'Id': '9710682', 'PubDate': '2022 Jan 25', 'EPubDate': '2022 Jan 25', 'Source': 'Bioinform Adv', 'AuthorList': ['van Gils JH', 'Gogishvili D', 'van Eck J', 'Bouwmeester R', 'van Dijk E', 'Abeln S'], 'Title': 'How sticky are our proteins? Quantifying hydrophobicity of the human proteome', 'Volume': '2', 'Issue': '1', 'Pages': 'vbac002', 'ArticleIds': {'pubmed': [], 'medline': [], 'pmid': '36699344', 'doi': '10.1093/bioadv/vbac002', 'pmcid': 'PMC9710682'}, 'DOI': '10.1093/bioadv/vbac002', 'FullJournalName': 'Bioinformatics Advances', 'SO': '2022 Jan 25;2(1):vbac002'}\n"]}]},{"cell_type":"code","source":["assemblies = Entrez.read(Entrez.esearch(db='assembly', term='GCA_003713205.1'))\n","assembly = assemblies['IdList'][0] if assemblies['IdList'] else None\n","print(f'assembly: {assembly}')\n","if assembly:\n","  summary = Entrez.read(Entrez.esummary(db='assembly', id=assembly))\n","  summary_data = summary['DocumentSummarySet']['DocumentSummary'][0]\n","  print(summary_data['Taxid'], summary_data['Organism'], summary_data['SpeciesTaxid'], summary_data['SpeciesName'], summary_data['AssemblyStatus'])\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"j0mKwf_QkCAg","executionInfo":{"status":"ok","timestamp":1675617976259,"user_tz":180,"elapsed":694,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"e2db70ed-4998-4e2f-c604-c0029eeb5a59"},"execution_count":53,"outputs":[{"output_type":"stream","name":"stdout","text":["assembly: 2070231\n","49369 Coffea eugenioides (eudicots) 49369 Coffea eugenioides Chromosome\n"]}]},{"cell_type":"markdown","source":["## Efetch\n","Retrieves records in the requested format from a list of one or more primary IDs.\n","https://www.ncbi.nlm.nih.gov/books/NBK25499/table/chapter4.T._valid_values_of__retmode_and/"],"metadata":{"id":"fmgy-Q0Fo3OV"}},{"cell_type":"code","source":["from Bio import Entrez, SeqIO\n","\n","handle = Entrez.esearch(db=\"nucleotide\",retmax=\"10\",term=\"S [Gene Name] AND Sars-Cov_2 [organism]\")\n","#handle = Entrez.esearch(db=\"nucleotide\",retmax=\"10\",term=\"S [Gene Name] AND Sars-Cov_2 [organism] AND Refseq\")\n","record = Entrez.read(handle)\n","print(record)\n","\n","# get first\n","summary = Entrez.read(Entrez.esummary(db='nucleotide', id=record[\"IdList\"][0]))[0]\n","print(summary)\n","\n","genbank_record = Entrez.efetch(db=\"nucleotide\", id=record[\"IdList\"][0],rettype=\"gb\", retmode=\"text\")\n","record = SeqIO.read(genbank_record, \"genbank\")\n","print(record)\n"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"YnY9U5WxtLMs","executionInfo":{"status":"ok","timestamp":1675620141902,"user_tz":180,"elapsed":1420,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"79e4839f-99dd-4056-81cf-a923c67ea5ae"},"execution_count":88,"outputs":[{"output_type":"stream","name":"stdout","text":["{'Count': '3337424', 'RetMax': '10', 'RetStart': '0', 'IdList': ['2438585312', '2438585299', '2438585287', '2438585274', '2438585261', '2438585249', '2438585236', '2438585223', '2438585210', '2438585198'], 'TranslationSet': [{'From': 'Sars-Cov_2[organism]', 'To': '\"Severe acute respiratory syndrome coronavirus 2\"[Organism]'}], 'TranslationStack': [{'Term': 'S[Gene Name]', 'Field': 'Gene Name', 'Count': '3398891', 'Explode': 'N'}, {'Term': '\"Severe acute respiratory syndrome coronavirus 2\"[Organism]', 'Field': 'Organism', 'Count': '6690456', 'Explode': 'Y'}, 'AND'], 'QueryTranslation': 'S[Gene Name] AND \"Severe acute respiratory syndrome coronavirus 2\"[Organism]'}\n","{'Item': [], 'Id': '2438585312', 'Caption': 'OQ365040', 'Title': 'Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/VA-CDC-LC0997448/2023 ORF1ab polyprotein (ORF1ab), ORF1a polyprotein (ORF1ab), surface glycoprotein (S), ORF3a protein (ORF3a), envelope protein (E), membrane glycoprotein (M), ORF6 protein (ORF6), ORF7a protein (ORF7a), and ORF7b (ORF7b) genes, complete cds; ORF8 gene, complete sequence; and nucleocapsid phosphoprotein (N) and ORF10 protein (ORF10) genes, complete cds', 'Extra': 'gi|2438585312|gb|OQ365040.1|[2438585312]', 'Gi': IntegerElement(2438585312, attributes={}), 'CreateDate': '2023/02/02', 'UpdateDate': '2023/02/02', 'Flags': IntegerElement(0, attributes={}), 'TaxId': IntegerElement(2697049, attributes={}), 'Length': IntegerElement(29660, attributes={}), 'Status': 'live', 'ReplacedBy': '', 'Comment': '  ', 'AccessionVersion': 'OQ365040.1'}\n","ID: OQ365040.1\n","Name: OQ365040\n","Description: Severe acute respiratory syndrome coronavirus 2 isolate SARS-CoV-2/human/USA/VA-CDC-LC0997448/2023 ORF1ab polyprotein (ORF1ab), ORF1a polyprotein (ORF1ab), surface glycoprotein (S), ORF3a protein (ORF3a), envelope protein (E), membrane glycoprotein (M), ORF6 protein (ORF6), ORF7a protein (ORF7a), and ORF7b (ORF7b) genes, complete cds; ORF8 gene, complete sequence; and nucleocapsid phosphoprotein (N) and ORF10 protein (ORF10) genes, complete cds\n","Database cross-references: BioProject:PRJNA716984, BioSample:SAMN33014276\n","Number of features: 55\n","/molecule_type=RNA\n","/topology=linear\n","/data_file_division=VRL\n","/date=02-FEB-2023\n","/accessions=['OQ365040']\n","/sequence_version=1\n","/keywords=['purposeofsampling:baselinesurveillance']\n","/source=Severe acute respiratory syndrome coronavirus 2 (SARS-CoV-2)\n","/organism=Severe acute respiratory syndrome coronavirus 2\n","/taxonomy=['Viruses', 'Riboviria', 'Orthornavirae', 'Pisuviricota', 'Pisoniviricetes', 'Nidovirales', 'Cornidovirineae', 'Coronaviridae', 'Orthocoronavirinae', 'Betacoronavirus', 'Sarbecovirus']\n","/references=[Reference(title='CDC Sars CoV2 Sequencing Baseline Constellation', ...), Reference(title='Direct Submission', ...)]\n","/structured_comment=defaultdict(<class 'dict'>, {'Assembly-Data': {'Assembly Method': 'CLC Genomics', 'Sequencing Technology': 'PacBio Sequel II'}})\n","Seq('CTGCATGCTTAGTGCACTCACGCAGTATAATTAATAACTAATTACTGTCGTTGA...TTA')\n"]}]},{"cell_type":"code","source":["from Bio import Entrez, SeqIO\n","# idtype: Specifies the type of identifier to return for sequence databases (nuccore, popset, protein). \n","# By default, ESearch returns GI numbers in its output. If idtype is set to ‘acc’, ESearch will return accession.version identifiers rather than GI numbers.\n","handle = Entrez.esearch(db=\"nuccore\",retmax=\"1\", term=\"P53 AND Homo Sapiens [organism]\", idtype='acc')\n","record = Entrez.read(handle)\n","print(record)\n","acc = record[\"IdList\"][0]\n","print('id', acc)\n","handle = Entrez.efetch(db=\"nucleotide\", id=record[\"IdList\"][0],rettype=\"gb\")\n","record = SeqIO.read(handle, \"genbank\")\n","print(record)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"4MwSI5guo2Av","executionInfo":{"status":"ok","timestamp":1675621089476,"user_tz":180,"elapsed":786,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"33af5cd8-244d-49f3-dc73-d3f40b225289"},"execution_count":116,"outputs":[{"output_type":"stream","name":"stdout","text":["{'Count': '7209', 'RetMax': '1', 'RetStart': '0', 'IdList': ['NG_033079.1'], 'TranslationSet': [{'From': 'Homo Sapiens[organism]', 'To': '\"Homo sapiens\"[Organism]'}], 'TranslationStack': [{'Term': 'P53[All Fields]', 'Field': 'All Fields', 'Count': '56999', 'Explode': 'N'}, {'Term': '\"Homo sapiens\"[Organism]', 'Field': 'Organism', 'Count': '28312156', 'Explode': 'Y'}, 'AND'], 'QueryTranslation': 'P53[All Fields] AND \"Homo sapiens\"[Organism]'}\n","id NG_033079.1\n","ID: NG_033079.1\n","Name: NG_033079\n","Description: Homo sapiens killin, p53 regulated DNA replication inhibitor (KLLN), RefSeqGene (LRG_1087) on chromosome 10\n","Number of features: 12\n","/molecule_type=DNA\n","/topology=linear\n","/data_file_division=PRI\n","/date=30-JAN-2023\n","/accessions=['NG_033079']\n","/sequence_version=1\n","/keywords=['RefSeq', 'RefSeqGene']\n","/source=Homo sapiens (human)\n","/organism=Homo sapiens\n","/taxonomy=['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Primates', 'Haplorrhini', 'Catarrhini', 'Hominidae', 'Homo']\n","/references=[Reference(title='Killin is a p53-regulated nuclear inhibitor of DNA synthesis', ...)]\n","/comment=REVIEWED REFSEQ: This record has been curated by NCBI staff in\n","collaboration with Sirisha Hesketh. The reference sequence was\n","derived from AC063965.8.\n","This sequence is a reference standard in the RefSeqGene project.\n","Summary: The protein encoded by this intronless gene is found in\n","the nucleus, where it can inhibit DNA synthesis and promote S phase\n","arrest coupled to apoptosis. The expression of this DNA binding\n","protein is upregulated by transcription factor p53. [provided by\n","RefSeq, Dec 2012].\n","Seq('TTTCTACAGAAAAAAAAAAATACATTTCTTAAGCTATGTTAATATTACATTTTC...GGT')\n"]}]},{"cell_type":"code","source":["fasta = SeqIO.read(Entrez.efetch(db=\"nucleotide\", id=acc, rettype=\"fasta\"), 'fasta')\n","fasta"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"1gOGBRmewErn","executionInfo":{"status":"ok","timestamp":1675621155682,"user_tz":180,"elapsed":554,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"e6a98580-ddbb-4c89-c93d-5e2b05bc6fd4"},"execution_count":120,"outputs":[{"output_type":"execute_result","data":{"text/plain":["SeqRecord(seq=Seq('TTTCTACAGAAAAAAAAAAATACATTTCTTAAGCTATGTTAATATTACATTTTC...GGT'), id='NG_033079.1', name='NG_033079.1', description='NG_033079.1 Homo sapiens killin, p53 regulated DNA replication inhibitor (KLLN), RefSeqGene (LRG_1087) on chromosome 10', dbxrefs=[])"]},"metadata":{},"execution_count":120}]},{"cell_type":"code","source":["from Bio import Entrez, SeqIO\n","handle = Entrez.efetch(db=\"nucleotide\", id=\"186972394\",rettype=\"gb\", retmode=\"text\")\n","record = SeqIO.read(handle, \"genbank\")\n","print(record)\n","handle.close()"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"cDLGnothsYYR","executionInfo":{"status":"ok","timestamp":1675619370187,"user_tz":180,"elapsed":326,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"a7997ab2-f552-43bc-db9f-88fe49453547"},"execution_count":72,"outputs":[{"output_type":"stream","name":"stdout","text":["ID: EU490707.1\n","Name: EU490707\n","Description: Selenipedium aequinoctiale maturase K (matK) gene, partial cds; chloroplast\n","Number of features: 3\n","/molecule_type=DNA\n","/topology=linear\n","/data_file_division=PLN\n","/date=26-JUL-2016\n","/accessions=['EU490707']\n","/sequence_version=1\n","/keywords=['']\n","/source=chloroplast Selenipedium aequinoctiale\n","/organism=Selenipedium aequinoctiale\n","/taxonomy=['Eukaryota', 'Viridiplantae', 'Streptophyta', 'Embryophyta', 'Tracheophyta', 'Spermatophyta', 'Magnoliopsida', 'Liliopsida', 'Asparagales', 'Orchidaceae', 'Cypripedioideae', 'Selenipedium']\n","/references=[Reference(title='Phylogenetic utility of ycf1 in orchids: a plastid gene more variable than matK', ...), Reference(title='Direct Submission', ...)]\n","Seq('ATTTTTTACGAACCTGTGGAAATTTTTGGTTATGACAATAAATCTAGTTTAGTA...GAA')\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"m8SDyoByv-9_"},"execution_count":null,"outputs":[]},{"cell_type":"markdown","source":["## NCBI Global Search\n","Busca global em todos os databases.\n","\n","\n"],"metadata":{"id":"i1lYkvAnnZ9R"}},{"cell_type":"code","source":["handle = Entrez.egquery(term=\"biopython\")\n","record = Entrez.read(handle)\n","handle.close()\n","for row in record[\"eGQueryResult\"]:\n","    if row[\"Count\"] != '0':\n","      print(row)"],"metadata":{"colab":{"base_uri":"https://localhost:8080/"},"id":"ZCh8ZLLsneiH","executionInfo":{"status":"ok","timestamp":1675618320598,"user_tz":180,"elapsed":434,"user":{"displayName":"Tania Mangolini","userId":"11639081150182463204"}},"outputId":"9c57d9c9-754c-453f-ef00-da955680c53a"},"execution_count":59,"outputs":[{"output_type":"stream","name":"stdout","text":["{'DbName': 'pubmed', 'MenuName': 'PubMed', 'Count': '41', 'Status': 'Ok'}\n","{'DbName': 'pmc', 'MenuName': 'PubMed Central', 'Count': '2715', 'Status': 'Ok'}\n","{'DbName': 'books', 'MenuName': 'Books', 'Count': '2', 'Status': 'Ok'}\n","{'DbName': 'pubmedhealth', 'MenuName': 'PubMed Health', 'Count': 'Error', 'Status': 'Database Error'}\n","{'DbName': 'ncbisearch', 'MenuName': 'Site Search', 'Count': '1', 'Status': 'Ok'}\n","{'DbName': 'nuccore', 'MenuName': 'Nucleotide', 'Count': '2', 'Status': 'Ok'}\n","{'DbName': 'sra', 'MenuName': 'SRA', 'Count': '1377', 'Status': 'Ok'}\n","{'DbName': 'biosystems', 'MenuName': 'BioSystems', 'Count': 'Error', 'Status': 'Database Error'}\n","{'DbName': 'gds', 'MenuName': 'GEO DataSets', 'Count': '17', 'Status': 'Ok'}\n","{'DbName': 'bioproject', 'MenuName': 'BioProject', 'Count': '1', 'Status': 'Ok'}\n"]}]},{"cell_type":"code","source":[],"metadata":{"id":"cmZjYiyqoqNy"},"execution_count":null,"outputs":[]}]}